scripts/globals/garrison.lua

```diff
@@ -596,46 +674,146 @@ xi.garrison.tick = function(npc)
+            zoneData.waveIndex = zoneData.waveIndex + 1
+            zoneData.groupIndex = 1
+            zoneData.nextSpawnTime = os.time() + xi.garrison.waves.delayBetweenGroups
+            zoneData.state = xi.garrison.state.BATTLE
+            zoneData.mobs = {}
+        end,
+
+        [xi.garrison.state.SPAWN_MOBS] = function()
+            debugLog("State: Spawn Mobs")
+            debugPrintToPlayers(zoneData, "Spawning mobs")
+
+            -- There are always at most 8 mobs + 1 boss for Garrison, so we will look up the
+            -- boss's ID using their name and then subtract 8 to get the starting mob ID.
+            local firstMobID = zone:queryEntitiesByName(zoneData.mobBoss)[1]:getID() - 8
+            local numMobs = xi.garrison.waves.mobsPerGroup
+            local poolSize = xi.garrison.waves.mobsPerGroup * xi.garrison.waves.groupsPerWave[zoneData.waveIndex]
+            local lastMobID = firstMobID + poolSize - 1
+
+            local mobIDs = pickMobsFromPool(firstMobID, lastMobID, numMobs, zoneData.mobs)
+
+            for _, mobID in ipairs(mobIDs) do
+                local mob = SpawnMob(mobID)
+                if mob ~= nil then
+                    addLevelCap(mob, zoneData.levelCap)
+                    mob:setRoamFlags(xi.roamFlag.SCRIPTED)
+                    table.insert(zoneData.mobs, mob:getID())
+                end
+            end
+
+            -- Once the mobs are spawned, make them aggro whatever NPCs are already up
+            aggroGroups(zoneData.mobs, zoneData.npcs)
+
+            zoneData.nextSpawnTime = os.time() + xi.garrison.waves.delayBetweenGroups
+            zoneData.state = xi.garrison.state.BATTLE
+            zoneData.groupIndex = zoneData.groupIndex + 1
+        end,
+
+        [xi.garrison.state.GRANT_LOOT] = function()
+            debugLog("State: Grant Loot")
+            debugPrintToPlayers(zoneData, "TODO: Loot")
+            zoneData.state = xi.garrison.state.ENDED
         end,
 
         [xi.garrison.state.ENDED] = function()
-            -- print("State: Ended")
+            debugLog("State: Ended")
+            -- TODO: Only check valid players
+            for _, entityId in pairs(zoneData.players) do
+                local entity = GetPlayerByID(entityId)
+                if entity ~= nil then
+                    entity:delStatusEffect(xi.effect.LEVEL_RESTRICTION)
+                end
+            end
+
+            for _, entityId in pairs(zoneData.npcs) do
+                DespawnMob(entityId)
+            end
+
+            for _, entityId in pairs(zoneData.mobs) do
+                DespawnMob(entityId)
+            end
+
             zoneData.continue = false
         end,
     }
 
-    -- TODO: Check to see if current wave is complete
-
-    -- TODO: Check to if all waves are completed
-
     if zoneData.continue then
-        npc:timer(2000, function(npcArg)
+        npc:timer(1000, function(npcArg)
```

Comment:
i think 2 seconds is a bit too slow since it does change how respawns happen, so I took it down to 1s